packages/node-patches/src/fs.ts

```diff
@@ -0,0 +1,458 @@
+/**
+ * @license
+ * Copyright 2019 The Bazel Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Stats, Dirent, Dir } from 'fs';
+import * as path from 'path';
+import * as util from 'util';
+// using require here on purpose so we can override methods with any
+// also even though imports are mutable in typescript the cognitive dissonance is too high because es modules
+const _fs = require('fs');
+
+//tslint:disable-next-line:no-any
+export const patcher = (fs: any = _fs, root: string) => {
+  fs = fs || _fs;
+  root = root || '';
+  if (!root) {
+    if (process.env.VERBOSE_LOGS) {
+      console.error('fs patcher called without root path ' + __filename);
+    }
+    return;
+  }
+
+  root = fs.realpathSync(root);
+
+  const origRealpath = fs.realpath.bind(fs);
+  const origLstat = fs.lstat.bind(fs);
+  const origStat = fs.stat.bind(fs);
+  const origStatSync = fs.statSync.bind(fs);
+  const origReadlink = fs.readlink.bind(fs);
+  const origLstatSync = fs.lstatSync.bind(fs);
+  const origRealpathSync = fs.realpathSync.bind(fs);
+  const origReadlinkSync = fs.readlinkSync.bind(fs);
+  const origReaddir = fs.readdir.bind(fs);
+  const origReaddirSync = fs.readdirSync.bind(fs);
+
+  const { isEscape, isOutPath } = escapeFunction(root);
+
+  const logged: { [k: string]: boolean } = {};
+
+  //tslint:disable-next-line:no-any
+  fs.lstat = (...args: any[]) => {
+    const ekey = new Error('').stack || '';
+    if (!logged[ekey]) {
+      logged[ekey] = true;
+    }
+
+    let cb = args.length > 1 ? args[args.length - 1] : undefined;
+    // preserve error when calling function without required callback.
+    if (cb) {
+      cb = once(cb);
+      args[args.length - 1] = (err: Error, stats: Stats) => {
+        if (err) return cb(err);
+
+        const linkPath = path.resolve(args[0]);
+        // if this is not a symlink or the path is not inside the root it has no way to escape.
+        if (!stats.isSymbolicLink() || !root || isOutPath(linkPath)) {
+          return cb(null, stats);
+        }
+
+        // this uses realpath here and this creates a divergence in behavior.
+        // the benefit is that if the
```

Comment:
This comment appears to have been cut off a little and continues down below

Reply:
i think i fixed the issue i was describing here when i moved to use lstat properly. thanks for the note! ill clean up